Main.hs

```diff
@@ -113,100 +113,169 @@ hsQName builtins f
+isSpecialTerm = show >>> \ case
+  "Haskell.Prelude.if_then_else_" -> Just ifThenElse
+  _ -> Nothing
+
+isSpecialCon :: QName -> Maybe (ConHead -> ConInfo -> Elims -> TCM (Hs.Exp ()))
+isSpecialCon = show >>> \ case
+  "Haskell.Prelude.Tuple._∷_" -> Just tupleTerm
+  _ -> Nothing
+
+isSpecialPat :: QName -> Maybe (ConHead -> ConPatternInfo -> [NamedArg DeBruijnPattern] -> TCM (Hs.Pat ()))
+isSpecialPat = show >>> \ case
+  "Haskell.Prelude.Tuple._∷_" -> Just tuplePat
+  _ -> Nothing
+
+isSpecialType :: QName -> Maybe (QName -> Elims -> TCM (Hs.Type ()))
+isSpecialType = show >>> \ case
+  "Haskell.Prelude.Tuple" -> Just tupleType
+  _ -> Nothing
+
+isSpecialName :: QName -> Maybe (Hs.QName ())
+isSpecialName = show >>> \ case
+    "Agda.Builtin.Nat.Nat"         -> unqual "Integer"
+    "Agda.Builtin.Float.Float"     -> unqual "Double"
+    "Agda.Builtin.Bool.Bool.false" -> unqual "False"
+    "Agda.Builtin.Bool.Bool.true"  -> unqual "True"
+    "Agda.Builtin.List.List"       -> special Hs.ListCon
+    "Agda.Builtin.List.List._∷_"   -> special Hs.Cons
+    "Agda.Builtin.List.List.[]"    -> special Hs.ListCon
+    "Agda.Builtin.Unit.⊤"          -> special Hs.UnitCon
+    "Agda.Builtin.Unit.tt"         -> special Hs.UnitCon
+    "Haskell.Prelude.Tuple.[]"     -> special Hs.UnitCon
+    _ -> Nothing
   where
-    unqual n  = Hs.UnQual () $ hsName n
-    special c = Hs.Special () $ c ()
+    unqual n  = Just $ Hs.UnQual () $ hsName n
+    special c = Just $ Hs.Special () $ c ()
+
+ifThenElse :: QName -> Elims -> TCM (Hs.Exp ())
+ifThenElse _ es = compileArgs es >>= \case
+  -- fully applied
+  (b : t : f : es') -> return $ Hs.If () b t f `eApp` es'
+  -- partially applied -> eta-expand
+  es' -> do
+    xs <- fmap Hs.name . drop (length es') <$> mapM freshString ["b", "t", "f"]
+    let [b, t, f] = es' ++ map Hs.var xs
+    return $ Hs.lamE (Hs.pvar <$> xs) $ Hs.If () b t f
+
+tupleType :: QName -> Elims -> TCM (Hs.Type ())
+tupleType _ es | Just [as] <- allApplyElims es = do
+  err <- sep [ text "Argument"
+             , nest 2 $ prettyTCM as
+             , text "to Tuple is not a concrete list" ]
+  xs <- makeList err (unArg as)
+  ts <- mapM compileType xs
+  return $ Hs.TyTuple () Hs.Boxed ts
+tupleType _ es =
+  genericDocError =<< text "Bad tuple arguments: " <?> prettyTCM es
+
+tupleTerm :: ConHead -> ConInfo -> Elims -> TCM (Hs.Exp ())
+tupleTerm cons i es = do
+  let v = Con cons i es
+  err <- sep [ text "Tuple value"
+             , nest 2 $ prettyTCM v
+             , text "does not have a known size." ]
+  xs <- makeList' "Haskell.Prelude.Tuple.[]" "Haskell.Prelude.Tuple._∷_" err v
+  ts <- mapM compileTerm xs
+  return $ Hs.Tuple () Hs.Boxed ts
+
+tuplePat :: ConHead -> ConPatternInfo -> [NamedArg DeBruijnPattern] -> TCM (Hs.Pat ())
+tuplePat cons i ps = do
+  let p = ConP cons i ps
+  err <- sep [ text "Tuple pattern"
+             , nest 2 $ prettyTCM p
+             , text "does not have a known size." ]
```

Comment:
Can we add an error-expectation test for this, or is it simply `__IMPOSSIBLE__`? (and similarly for `tupleTerm`)

Reply:
It's not impossible. We haven't set up any negative tests yet, but this is definitely a candidate. Something like
```agda
bla : a -> Tuple as -> Tuple (a :: as)
bla x xs = x :: xs
```
Might be tricky to not hit the error in the type first though.

Reply:
Note: an artificial example for getting to this error:
```agda
bla : Tuple [] → Tuple (Nat ∷ [])
bla xs = 1 ∷ xs
{-# COMPILE AGDA2HS bla #-}

```